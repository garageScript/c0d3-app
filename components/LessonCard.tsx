import React from 'react'
import { useQuery } from '@apollo/react-hooks'
import { CheckCircle } from 'react-feather'
import { GET_SUBMISSIONS } from '../graphql/queries'
import {
  Props,
  ReviewButtonProps,
  ReviewCountProps
} from '../@types/lessonCard'
import NavLink from './NavLink'
import '../scss/lessonCard.scss'

const ReviewCount: React.FC<ReviewCountProps> = props => {
  if (props.shouldNotGetCount) {
    return null
  }

  const { loading, data } = useQuery(GET_SUBMISSIONS, {
    variables: {
      in: {
        id: `${props.lessonId}`
      }
    }
  })

  if (loading) {
    return (
      <div className="spinner-border spinner-border-sm mx-1" role="status" />
    )
  }

  if (!data) {
    return null
  }
  const pendingSubmissionsCount = data.submissions.reduce(
    (acc: number, val: any) => {
      if (val.status === 'open') {
        acc = acc + 1
        return acc
      }
      return acc
    },
    0
  )

  return <span> {pendingSubmissionsCount} </span>
}

const ReviewButton: React.FC<ReviewButtonProps> = props => {
  if (!props.isCompleted) {
    return null
  }
  return (
    <a
      href={props.reviewUrl}
      className="btn btn-sm bg-primary text-white float-right mb-2 mr-2"
    >
      Review
      <ReviewCount
        shouldNotGetCount={props.shouldNotGetCount}
        lessonId={props.lessonId}
      />
      Submissions
    </a>
  )
}

const LessonCard: React.FC<Props> = props => {
  const containerClass =
    props.currentState === 'inProgress'
      ? 'lesson-card__container_inprogress border-primary'
      : 'border-0'

  return (
    <div className={`card shadow-sm mt-3 ${containerClass}`}>
      <div className="d-flex p-2">
        <img
          src={`/curriculumAssets/lessonCoversSvg/${props.coverImg}`}
          alt={props.coverImg}
        />

        <div className="w-100 pl-4">
          {props.currentState === 'completed' && (
            <span className="badge badge-pill badge-success float-right mt-2 mr-2 p-2 d-flex align-items-center">
              <CheckCircle style={{ height: '15px' }} />
              <span className="mx-1">COMPLETED</span>
            </span>
          )}
          <div>
            <h4 className="lesson-card__title font-weight-bold mt-3">
              <NavLink
                as={`/curriculum/${props.lessonId}`}
                path="/curriculum/[lesson]"
                text={props.title}
              />
            </h4>
            <div>
              <div className="d-inline-block mr-4">
                <img
                  className="mr-2"
                  src="/curriculumAssets/icons/icon-challenge.svg"
                  alt="icon-challenge"
                />
                <span className="lesson-card__icon-text">
                  {props.challengeCount} CHALLENGES
                </span>
              </div>
            </div>
            <p className="lesson-card__description mt-2">{props.description}</p>
          </div>
          <ReviewButton
            isCompleted={props.currentState === 'completed'}
            reviewUrl={props.reviewUrl}
            shouldNotGetCount={props.shouldNotGetCount}
            lessonId={props.lessonId}
          />
        </div>
      </div>

      {props.currentState === 'inProgress' && (
        <div className="p-2 bg-primary">
          <NavLink
            path={props.docUrl}
            text="Start Lesson"
            className="lesson-card__button btn btn-light mr-2 my-1 text-primary"
            external
          />
          <NavLink
            path={props.challengesUrl}
            text="View Challenges"
            className="lesson-card__button btn bg-primary my-1 text-white border border-white"
          />
        </div>
      )}
    </div>
  )
}

export default LessonCard
