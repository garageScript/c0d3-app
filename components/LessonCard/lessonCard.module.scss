@use '../../scss/variables.module';

.lesson-card__container_inprogress {
  border-width: 2px;
}
.lesson-card__title {
  font-size: 1.5rem;
  line-height: 1.3;
  margin-bottom: 0.75rem;
}

.lesson-card__badge {
  align-self: flex-start;
  padding-left: 1em;
  padding-right: 1em;
  margin-left: auto;
  background-color: variables.$success;
  @media (max-width: 768px) {
    background-color: transparent;
    color: #00cf92;
    margin-left: auto;
    & > svg {
      width: 30px;
      height: 30px;
      stroke-width: 3px;
    }
  }
}

.lesson-card__icon-text {
  font-size: 0.8rem;
  line-height: 1.4;
}
.lesson-card__button {
  font-size: 13px;
  min-width: 134px;
}
.lesson-card__image {
  object-fit: contain;
}
.lesson-card__description {
  flex: 1;
}
