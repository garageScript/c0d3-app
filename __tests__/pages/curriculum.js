jest.mock('@apollo/react-hooks')
jest.mock('../../components/LessonCard')
import React from 'react'
import { useQuery } from '@apollo/react-hooks'
import Curriculum from '../../pages/curriculum'
import { render } from '@testing-library/react'
import LessonCard from '../../components/LessonCard'

describe('Curriculum Page', () => {
  test('Should render loading when loading', async () => {
    useQuery.mockReturnValue({
      loading: true
    })

    const { container } = render(<Curriculum />)
    expect(container).toMatchSnapshot()
  })
  test('Should render ellpsis when no data is present', async () => {
    useQuery.mockReturnValue({
      loading: false
    })

    const { container } = render(<Curriculum />)
    expect(container).toMatchSnapshot()
  })
  test('Should render data with lesson state completed', async () => {
    useQuery.mockReturnValue({
      data: {
        curriculumStatus: [
          {
            id: 4,
            title: 'lesson 4',
            description: '4th lesson',
            order: 4,
            challenges: [
              {
                id: 4
              }
            ],
            currentUser: {
              userLesson: {}
            },
            docUrl: 'lesson4.com'
          },
          {
            id: 1,
            title: 'lesson 1',
            description: '1st lesson',
            order: 1,
            challenges: [
              {
                id: 1
              }
            ],
            currentUser: {
              userLesson: {}
            },
            docUrl: 'lesson1.com'
          }
        ]
      }
    })
    LessonCard.mockReturnValue(<h1>LessonCard</h1>)

    const { container } = render(<Curriculum />)
    expect(container).toMatchSnapshot()
  })
  test('Should render data with lesson state inProgress', async () => {
    useQuery.mockReturnValue({
      data: {
        curriculumStatus: [
          {
            id: 4,
            title: 'lesson 4',
            description: '4th lesson',
            order: 4,
            challenges: [
              {
                id: 4
              }
            ],
            currentUser: {
              userLesson: {
                isPassed: '234234'
              }
            },
            docUrl: 'lesson4.com'
          },
          {
            id: 1,
            title: 'lesson 1',
            description: '1st lesson',
            order: 1,
            challenges: [
              {
                id: 1
              }
            ],
            currentUser: {
              userLesson: {
                isPassed: '9878990'
              }
            },
            docUrl: 'lesson1.com'
          }
        ]
      }
    })
    LessonCard.mockReturnValue(<h1>LessonCard</h1>)

    const { container } = render(<Curriculum />)
    expect(container).toMatchSnapshot()
  })
})
