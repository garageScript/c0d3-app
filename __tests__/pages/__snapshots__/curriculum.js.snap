// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Curriculum Page Should render data with lesson state completed 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link active"
              href="/"
            >
              <span
                class="sr-only"
              >
                (current)
              </span>
              Home
            </a>
            <a
              class="nav-item nav-link"
              href="/#learning"
            >
              Learning Process
            </a>
            <a
              class="nav-item nav-link"
              href="https://c0d3.com/book"
              rel="noopener noreferrer"
              target="_blank"
            >
              Resources
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <div
      class="row mt-4"
    >
      <div
        class="col-8"
      >
        <h1>
          LessonCard
        </h1>
        <h1>
          LessonCard
        </h1>
      </div>
      <div
        class="col-4"
      >
        <div
          class="progress-card__container d-flex card shadow-sm mt-3 bg-primary text-white p-2 border-0"
        >
          <div
            class="card-body"
          >
            <div
              class="progress-card__progress-bar d-flex justify-content-center align-items-center rounded-circle"
            >
              <span
                class="font-weight-bold"
              >
                0
                %
              </span>
            </div>
            <h4
              class="progress-card__title mt-3"
            >
              You're off to a great start!
            </h4>
            <div
              class="mt-3"
            >
              <p>
                Start by setting up your development environment and then follow the lessons in the course.
              </p>
            </div>
            <a
              class="btn btn-light progress-card__button mt-2 text-primary"
              href="https://www.notion.so/Setup-Instructions-fc8f8fcc1376482ead839fa6b1034cb4#10ab6286f4d942c58bd330f68cb8cb95"
              rel="noopener noreferrer"
              target="_blank"
            >
              Setup Development Environment
            </a>
          </div>
        </div>
        <div
          class="announcement-card__container card shadow-sm mt-3 p-2 d-flex border-0"
        >
          <div
            class="card-body"
          >
            <h1
              class="announcement-card__title"
            >
              General Announcements
            </h1>
            <div
              class="mt-3"
            >
              <p
                class="font-weight-bold"
              >
                Updates and Guidelines
              </p>
              <p
                class="mt-4"
              >
                To make space for other students on our servers, your account will be deleted after 30 days of inactivity.
              </p>
              <p
                class="mt-4"
              >
                Take each lesson challenge seriously and do them over and over again until you can solve them. With the exception End to End, all challenges are questions and exercises taken from real interviews.
              </p>
              <p
                class="mt-4"
              >
                This lesson will not only prepare you for interviews, but it will also help teach you the skills that you need to become an effective engineer.
              </p>
              <p
                class="mt-4"
              >
                After completing Foundations of JavaScript, Variables & Functions, Array, Objects, End to End, HTML/CSS/JavaScript, React/GraphQL/SocketIO, you will be technically ready to contribute to our codebase.
              </p>
            </div>
          </div>
        </div>
        <div
          class="additional-resources__container mt-4 mb-0"
        >
          <h1
            class="additional-resources__title text-muted"
          >
            ADDITIONAL RESOURCES
          </h1>
          <hr
            class="my-2"
          />
          <a
            class="mx-3"
            href="https://www.notion.so/Setup-Instructions-fc8f8fcc1376482ead839fa6b1034cb4"
            rel="noopener noreferrer"
            target="_blank"
          >
            Setup Instructions
          </a>
          <hr
            class="my-2"
          />
          <a
            class="mx-3"
            href="#"
          >
            Github Basics
          </a>
          <hr
            class="my-2"
          />
          <a
            class="mx-3"
            href="https://stackoverflow.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Stack Overflow
          </a>
          <hr
            class="my-2"
          />
          <a
            class="mx-3"
            href="https://chat.c0d3.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Help & Support
          </a>
          <hr
            class="my-2"
          />
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    Â© Copyright 2020 GarageScript
    <br />
    All Rights Reserved
  </footer>
</div>
`;

exports[`Curriculum Page Should render data with lesson state inProgress 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link active"
              href="/"
            >
              <span
                class="sr-only"
              >
                (current)
              </span>
              Home
            </a>
            <a
              class="nav-item nav-link"
              href="/#learning"
            >
              Learning Process
            </a>
            <a
              class="nav-item nav-link"
              href="https://c0d3.com/book"
              rel="noopener noreferrer"
              target="_blank"
            >
              Resources
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <div
      class="row mt-4"
    >
      <div
        class="col-8"
      >
        <h1>
          LessonCard
        </h1>
        <h1>
          LessonCard
        </h1>
      </div>
      <div
        class="col-4"
      >
        <div
          class="progress-card__container d-flex card shadow-sm mt-3 bg-primary text-white p-2 border-0"
        >
          <div
            class="card-body"
          >
            <div
              class="progress-card__progress-bar d-flex justify-content-center align-items-center rounded-circle"
            >
              <span
                class="font-weight-bold"
              >
                0
                %
              </span>
            </div>
            <h4
              class="progress-card__title mt-3"
            >
              You're off to a great start!
            </h4>
            <div
              class="mt-3"
            >
              <p>
                Start by setting up your development environment and then follow the lessons in the course.
              </p>
            </div>
            <a
              class="btn btn-light progress-card__button mt-2 text-primary"
              href="https://www.notion.so/Setup-Instructions-fc8f8fcc1376482ead839fa6b1034cb4#10ab6286f4d942c58bd330f68cb8cb95"
              rel="noopener noreferrer"
              target="_blank"
            >
              Setup Development Environment
            </a>
          </div>
        </div>
        <div
          class="announcement-card__container card shadow-sm mt-3 p-2 d-flex border-0"
        >
          <div
            class="card-body"
          >
            <h1
              class="announcement-card__title"
            >
              General Announcements
            </h1>
            <div
              class="mt-3"
            >
              <p
                class="font-weight-bold"
              >
                Updates and Guidelines
              </p>
              <p
                class="mt-4"
              >
                To make space for other students on our servers, your account will be deleted after 30 days of inactivity.
              </p>
              <p
                class="mt-4"
              >
                Take each lesson challenge seriously and do them over and over again until you can solve them. With the exception End to End, all challenges are questions and exercises taken from real interviews.
              </p>
              <p
                class="mt-4"
              >
                This lesson will not only prepare you for interviews, but it will also help teach you the skills that you need to become an effective engineer.
              </p>
              <p
                class="mt-4"
              >
                After completing Foundations of JavaScript, Variables & Functions, Array, Objects, End to End, HTML/CSS/JavaScript, React/GraphQL/SocketIO, you will be technically ready to contribute to our codebase.
              </p>
            </div>
          </div>
        </div>
        <div
          class="additional-resources__container mt-4 mb-0"
        >
          <h1
            class="additional-resources__title text-muted"
          >
            ADDITIONAL RESOURCES
          </h1>
          <hr
            class="my-2"
          />
          <a
            class="mx-3"
            href="https://www.notion.so/Setup-Instructions-fc8f8fcc1376482ead839fa6b1034cb4"
            rel="noopener noreferrer"
            target="_blank"
          >
            Setup Instructions
          </a>
          <hr
            class="my-2"
          />
          <a
            class="mx-3"
            href="#"
          >
            Github Basics
          </a>
          <hr
            class="my-2"
          />
          <a
            class="mx-3"
            href="https://stackoverflow.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Stack Overflow
          </a>
          <hr
            class="my-2"
          />
          <a
            class="mx-3"
            href="https://chat.c0d3.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Help & Support
          </a>
          <hr
            class="my-2"
          />
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    Â© Copyright 2020 GarageScript
    <br />
    All Rights Reserved
  </footer>
</div>
`;

exports[`Curriculum Page Should render ellpsis when no data is present 1`] = `
<div>
  <h1>
    No Data...
  </h1>
</div>
`;

exports[`Curriculum Page Should render loading when loading 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link active"
              href="/"
            >
              <span
                class="sr-only"
              >
                (current)
              </span>
              Home
            </a>
            <a
              class="nav-item nav-link"
              href="/#learning"
            >
              Learning Process
            </a>
            <a
              class="nav-item nav-link"
              href="https://c0d3.com/book"
              rel="noopener noreferrer"
              target="_blank"
            >
              Resources
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <div
      class="mt-4 card border-0 shadow-sm"
    >
      <div
        class="card-body d-flex flex-column align-items-center"
      >
        <h1>
          Loading
        </h1>
        <div
          class="spinner-border text-primary mt-1"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    Â© Copyright 2020 GarageScript
    <br />
    All Rights Reserved
  </footer>
</div>
`;
